### PR TITLE
Add Python conformance tests

### PR DESCRIPTION
**Pull request by tothtamas28**
Ports conformance tests to Python. Also changes the `test-conformance` target to run the Python tests.

### CODE REVIEW

**Comment by tothtamas28** [comment]
Tried not parsing the KORE text output, but that does not seem to have a significant performance impact (on my machine, both versions took 12 minutes).

**Comment by ehildenb** [comment]
I ran this locally:

```
make -C kevm-pyk test-integration TEST_ARGS='-n24 -k test_conformance' &> output
```

And got this error:

```
==================================== ERRORS ====================================
__________ ERROR collecting src/tests/integration/test_conformance.py __________
src/tests/integration/test_conformance.py:13: in <module>
    from kevm_pyk.interpreter import interpret
src/kevm_pyk/interpreter.py:8: in <module>
    from . import config
src/kevm_pyk/config.py:38: in <module>
    KEVM_LIB: Final = _kevm_lib()
src/kevm_pyk/config.py:20: in _kevm_lib
    check_dir_path(kevm_lib)
../../../.cache/pypoetry/virtualenvs/kevm-pyk-oCcYELil-py3.10/lib/python3.10/site-packages/pyk/cli_utils.py:38: in check_dir_path
    raise ValueError(f'Directory does not exist: {path}')
E   ValueError: Directory does not exist: /home/dev/src/evm-semantics-pr/kevm-pyk/.build/usr/lib/kevm
=========================== short test summary info ============================
ERROR src/tests/integration/test_conformance.py - ValueError: Directory does ...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!! xdist.dsession.Interrupted: stopping after 1 failures !!!!!!!!!!!!!
=============================== 1 error in 1.20s ===============================
make: *** [Makefile:36: test-integration] Error 2
make: Leaving directory '/home/dev/src/evm-semantics-pr/kevm-pyk'
```

It is somehow looking in the wrong directory there (notice the extra `kevm-pyk` before the `.build`). Do I need to manually do `KEVM_LIB=...` before running the command?

**Comment by tothtamas28** [comment]
Folder `.build/usr/lib/kevm` should exist after `make build` I think. Or is this on a Nix build?

**Comment by ehildenb** [comment]
Notice the extra `kevm-pyk` in the PATH. The folder `.build/usr/lib/kevm` does exist, but `kevm-pyk/.build/usr/lib/kevm` (the place where it's looking), does not.

**Comment by ehildenb** [comment]
Calling this instead works:

```
KEVM_LIB=$(pwd)/.build/usr/lib/kevm make -C kevm-pyk test-integration TEST_ARGS='-n24 -k test_conformance' &> output
```

**Comment by ehildenb** [comment]
Ah, indeed that is it.

**Review by anvacaru — APPROVED** [review]

**Comment by tothtamas28** [comment]
> Are we able to run an individual test somehow, by passing extra arguments to the test harness?

Yes, `pytest -k` allows that.

Selecting a single test:

```
$ make -C kevm-pyk test-integration TEST_ARGS+='-n0 --collect-only -k vmLogTest/log4_logMemsize
Zero.json'
...
<Package integration>
  <Module test_conformance.py>
    <Function test_vm[vmLogTest/log4_logMemsizeZero.json]>

```

All `VmLogTest`s:

```
make -C kevm-pyk test-integration TEST_ARGS+='-n0 --collect-only -k vmLogTest'
...
<Package integration>
  <Module test_conformance.py>
    <Function test_vm[vmLogTest/log4_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log3_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log2_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log2_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log4_Caller.json]>
    <Function test_vm[vmLogTest/log1_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log2_Caller.json]>
    <Function test_vm[vmLogTest/log1_MaxTopic.json]>
    <Function test_vm[vmLogTest/log4_PC.json]>
    <Function test_vm[vmLogTest/log0_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log4_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log4_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log3_emptyMem.json]>
    <Function test_vm[vmLogTest/log3_MaxTopic.json]>
    <Function test_vm[vmLogTest/log3_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log3_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log2_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log1_emptyMem.json]>
    <Function test_vm[vmLogTest/log0_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log3_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log0_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log0_emptyMem.json]>
    <Function test_vm[vmLogTest/log1_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log3_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log1_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log0_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log_2logs.json]>
    <Function test_vm[vmLogTest/log2_emptyMem.json]>
    <Function test_vm[vmLogTest/log2_MaxTopic.json]>
    <Function test_vm[vmLogTest/log3_PC.json]>
    <Function test_vm[vmLogTest/log4_emptyMem.json]>
    <Function test_vm[vmLogTest/log1_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log3_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log0_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log4_MaxTopic.json]>
    <Function test_vm[vmLogTest/log0_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log4_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log2_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log2_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log1_Caller.json]>
    <Function test_vm[vmLogTest/log3_Caller.json]>
    <Function test_vm[vmLogTest/log1_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log2_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log1_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log4_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log4_logMemsizeZero.json]>
```

All `VmLogTest`s, except `log4` ones:

```
$ make -C kevm-pyk test-integration TEST_ARGS+='-n0 --collect-only -k "vmLogTest and not log4"'
...
<Package integration>
  <Module test_conformance.py>
    <Function test_vm[vmLogTest/log3_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log2_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log2_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log1_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log2_Caller.json]>
    <Function test_vm[vmLogTest/log1_MaxTopic.json]>
    <Function test_vm[vmLogTest/log0_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log3_emptyMem.json]>
    <Function test_vm[vmLogTest/log3_MaxTopic.json]>
    <Function test_vm[vmLogTest/log3_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log3_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log2_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log1_emptyMem.json]>
    <Function test_vm[vmLogTest/log0_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log3_nonEmptyMem.json]>
    <Function test_vm[vmLogTest/log0_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log0_emptyMem.json]>
    <Function test_vm[vmLogTest/log1_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log3_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log1_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log0_logMemsizeZero.json]>
    <Function test_vm[vmLogTest/log_2logs.json]>
    <Function test_vm[vmLogTest/log2_emptyMem.json]>
    <Function test_vm[vmLogTest/log2_MaxTopic.json]>
    <Function test_vm[vmLogTest/log3_PC.json]>
    <Function test_vm[vmLogTest/log1_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log3_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log0_logMemStartTooHigh.json]>
    <Function test_vm[vmLogTest/log0_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log2_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log2_nonEmptyMem_logMemSize1_logMemStart31.json]>
    <Function test_vm[vmLogTest/log1_Caller.json]>
    <Function test_vm[vmLogTest/log3_Caller.json]>
    <Function test_vm[vmLogTest/log1_nonEmptyMem_logMemSize1.json]>
    <Function test_vm[vmLogTest/log2_logMemsizeTooHigh.json]>
    <Function test_vm[vmLogTest/log1_logMemsizeTooHigh.json]>
```

**Comment by tothtamas28** [comment]
> When we manually mangle a test (by changing something in one of teh `post` blocks of the test, for example), does the overall test harness fail?

For a `bchain` test I was able to confirm this  by modifying `postState`.

For a `vmtest` modifying `post` didn't make the test fail though, but then the `make` target didn't fail either. Is this expected?

File: kevm-pyk/src/kevm_pyk/interpreter.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    from typing import Any
+
+    from pyk.kore.syntax import Pattern
+
+
+def interpret(gst_data: Any, schedule: str, mode: str, chainid: int, *, check: bool = True) -> Pattern:
+    proc_res = _interpret(gst_data, schedule, mode, chainid)
+
+    if check:
+        proc_res.check_returncode()
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/interpreter.py:22]: Yes, it should be the case.

[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/interpreter.py:22]: The generated `interpreter` binary should exit with the contents of the `<exit-code>` cell.

If we're directly calling via bindings, then I guess we have to read that cell ourselves.

File: Makefile [hunk 2]
```diff
@@ -625,7 +626,8 @@ rest_vm_tests    = $(filter-out $(passing_vm_tests), $(all_vm_tests))
 
 test-all-vm: $(all_vm_tests:=.run)
 test-rest-vm: $(rest_vm_tests:=.run)
-test-vm: $(passing_vm_tests:=.run)
+test-vm: poetry build-kevm build-llvm
+	$(MAKE) -C kevm-pyk/ test-integration TEST_ARGS+='-k test_vm -n8'
 
 all_bchain_tests     = $(wildcard tests/ethereum-tests/BlockchainTests/GeneralStateTests/*/*.json)                            \
                        $(wildcard tests/ethereum-tests/LegacyTests/Constantinople/BlockchainTests/GeneralStateTests/*/*.json)
```
Comment (ehildenb) [inline Makefile:633]: Do we need to keep all the logic which enumerates the tests in the Makefile anymore?

Comment (tothtamas28) [reply Makefile]: Good point. A few targets, namely `test-all-vm`, `test-rest-vm`, `test-all-bchain`, `test-rest-bchain`, were not covered by the Python test harness. I added test functions accordingly and cleaned up the `Makefile`.

**Pull request merged**